### PR TITLE
Refactor recognition part

### PR DESCRIPTION
**Pull request by kan-bayashi**
#334 

- unified n-best processing 
- added score to json

Now recog part is the same among chainer and pytorch.

### CODE REVIEW

**Comment by kan-bayashi** [comment]
Done! Looks clean!

File: src/asr/asr_chainer.py [hunk 1]
```diff
@@ -470,56 +471,46 @@ def recog(args):
...
-                new_json[name]['rec_text' + '[' + '{:05d}'.format(i) + ']'] = seq_hat_text
-                new_json[name]['score' + '[' + '{:05d}'.format(i) + ']'] = hyp['score']
+    with chainer.no_backprop_mode():
+        for name in js.keys():
+            # perform decoding and get n-best results
+            logging.info('decoding ' + name)
+            feat = kaldi_io_py.read_mat(js[name]['input'][0]['feat'])
+            nbest_hyps = e2e.recognize(feat, args, train_args.char_list, rnnlm)
+
+            # get ground-truth
```
Comment (sw005320) [inline src/asr/asr_chainer.py]: You could also make a new function for these lines (484--513) (e.g., `new_json[name] = add_results_to_json(js[name], nbest_hyps)`) used for both asr_chainer.py and asr_pytorch.py

**Comment by kan-bayashi** [comment]
N-best result sample
```
        "4k2c0308": {
            "output": [
                {
                    "name": "target1",
                    "rec_text": "OF COURSE THERE ISN'T ANY GUARANTEE THE COMPANY WILL KEEP IT THOUGH HAND<eos>",
                    "rec_token": "O F <space> C O U R S E <space> T H E R E <space> I S N ' T <space> A N Y <space> G U A R A N T E E <space> T H E <space> C O M P A N Y <space> W I L L <space> K E E P <space> I T <space> T H O U G H <space> H A N D <eos>",
                    "rec_tokenid": "34 25 18 22 34 40 37 38 24 18 39 27 24 37 24 18 28 38 33 5 39 18 20 33 44 18 26 40 20 37 20 33 39 24 24 18 39 27 24 18 22 34 32 35 20 33 44 18 42 28 31 31 18 30 24 24 35 18 28 39 18 39 27 34 40 26 27 18 27 20 33 23 51",
                    "score": -44.3232421875,
                    "shape": [
                        72,
                        52
                    ],
                    "text": "OF COURSE THERE ISN'T ANY GUARANTEE THE COMPANY WILL KEEP ITS HOT HAND <NOISE>",
                    "token": "O F <space> C O U R S E <space> T H E R E <space> I S N ' T <space> A N Y <space> G U A R A N T E E <space> T H E <space> C O M P A N Y <space> W I L L <space> K E E P <space> I T S <space> H O T <space> H A N D <space> <NOISE>",
                    "tokenid": "34 25 18 22 34 40 37 38 24 18 39 27 24 37 24 18 28 38 33 5 39 18 20 33 44 18 26 40 20 37 20 33 39 24 24 18 39 27 24 18 22 34 32 35 20 33 44 18 42 28 31 31 18 30 24 24 35 18 28 39 38 18 27 34 39 18 27 20 33 23 18 17"
                },
                {
                    "name": "target2",
                    "rec_text": "OF COURSE THERE ISN'T ANY GUARANTEE THE COMPANY WILL KEEP IT THOUGHT HAND<eos>",
                    "rec_token": "O F <space> C O U R S E <space> T H E R E <space> I S N ' T <space> A N Y <space> G U A R A N T E E <space> T H E <space> C O M P A N Y <space> W I L L <space> K E E P <space> I T <space> T H O U G H T <space> H A N D <eos>",
                    "rec_tokenid": "34 25 18 22 34 40 37 38 24 18 39 27 24 37 24 18 28 38 33 5 39 18 20 33 44 18 26 40 20 37 20 33 39 24 24 18 39 27 24 18 22 34 32 35 20 33 44 18 42 28 31 31 18 30 24 24 35 18 28 39 18 39 27 34 40 26 27 39 18 27 20 33 23 51",
                    "score": -45.5652961730957,
                    "shape": [
                        72,
                        52
                    ],
                    "text": "OF COURSE THERE ISN'T ANY GUARANTEE THE COMPANY WILL KEEP ITS HOT HAND <NOISE>",
                    "token": "O F <space> C O U R S E <space> T H E R E <space> I S N ' T <space> A N Y <space> G U A R A N T E E <space> T H E <space> C O M P A N Y <space> W I L L <space> K E E P <space> I T S <space> H O T <space> H A N D <space> <NOISE>",
                    "tokenid": "34 25 18 22 34 40 37 38 24 18 39 27 24 37 24 18 28 38 33 5 39 18 20 33 44 18 26 40 20 37 20 33 39 24 24 18 39 27 24 18 22 34 32 35 20 33 44 18 42 28 31 31 18 30 24 24 35 18 28 39 38 18 27 34 39 18 27 20 33 23 18 17"
                },
                {
                    "name": "target3",
                    "rec_text": "OF COURSE THERE ISN'T ANY GUARANTEE THE COMPANY WILL KEEP IT THOUGHT HANDS<eos>",
                    "rec_token": "O F <space> C O U R S E <space> T H E R E <space> I S N ' T <space> A N Y <space> G U A R A N T E E <space> T H E <space> C O M P A N Y <space> W I L L <space> K E E P <space> I T <space> T H O U G H T <space> H A N D S <eos>",
                    "rec_tokenid": "34 25 18 22 34 40 37 38 24 18 39 27 24 37 24 18 28 38 33 5 39 18 20 33 44 18 26 40 20 37 20 33 39 24 24 18 39 27 24 18 22 34 32 35 20 33 44 18 42 28 31 31 18 30 24 24 35 18 28 39 18 39 27 34 40 26 27 39 18 27 20 33 23 38 51",
                    "score": -53.03976821899414,
                    "shape": [
                        72,
                        52
                    ],
                    "text": "OF COURSE THERE ISN'T ANY GUARANTEE THE COMPANY WILL KEEP ITS HOT HAND <NOISE>",
                    "token": "O F <space> C O U R S E <space> T H E R E <space> I S N ' T <space> A N Y <space> G U A R A N T E E <space> T H E <space> C O M P A N Y <space> W I L L <space> K E E P <space> I T S <space> H O T <space> H A N D <space> <NOISE>",
                    "tokenid": "34 25 18 22 34 40 37 38 24 18 39 27 24 37 24 18 28 38 33 5 39 18 20 33 44 18 26 40 20 37 20 33 39 24 24 18 39 27 24 18 22 34 32 35 20 33 44 18 42 28 31 31 18 30 24 24 35 18 28 39 38 18 27 34 39 18 27 20 33 23 18 17"
                }
            ],
```

**Comment by sw005320** [comment]
Thanks.
The current N-best result sample may have some conflicts when we use several targets (e.g., target1 is an English text and target2 is a German text).
How about using `"name": "target1[%d]"`?.
Then, we need to modify the scoring script as well.

**Comment by kan-bayashi** [comment]
Changed format

```
"4k2c0308": {
    "output": [
        {
            "name": "target1[1]",
            "rec_text": "OF COURSE THERE ISN'T ANY GUARANTEE THE COMPANY WILL KEEP ITS TOUGH HAND<eos>",
            "rec_token": "O F <space> C O U R S E <space> T H E R E <space> I S N ' T <space> A N Y <space> G U A R A N T E E <space> T H E <space> C O M P A N Y <space> W I L L <space> K E E P <space> I T S <space> T O U G H <space> H A N D <eos>",
            "rec_tokenid": "34 25 18 22 34 40 37 38 24 18 39 27 24 37 24 18 28 38 33 5 39 18 20 33 44 18 26 40 20 37 20 33 39 24 24 18 39 27 24 18 22 34 32 35 20 33 44 18 42 28 31 31 18 30 24 24 35 18 28 39 38 18 39 34 40 26 27 18 27 20 33 23 51",
            "score": -64.63552856445312,
            "shape": [
                72,
                52
            ],
            "text": "OF COURSE THERE ISN'T ANY GUARANTEE THE COMPANY WILL KEEP ITS HOT HAND <NOISE>",
            "token": "O F <space> C O U R S E <space> T H E R E <space> I S N ' T <space> A N Y <space> G U A R A N T E E <space> T H E <space> C O M P A N Y <space> W I L L <space> K E E P <space> I T S <space> H O T <space> H A N D <space> <NOISE>",
            "tokenid": "34 25 18 22 34 40 37 38 24 18 39 27 24 37 24 18 28 38 33 5 39 18 20 33 44 18 26 40 20 37 20 33 39 24 24 18 39 27 24 18 22 34 32 35 20 33 44 18 42 28 31 31 18 30 24 24 35 18 28 39 38 18 27 34 39 18 27 20 33 23 18 17"
        },
            "name": "target1[2]",
            "rec_text": "OF COURSE THERE ISN'T ANY GUARANTEE THE COMPANY WILL KEEP ITS THOUGH HAND<eos>",
            "rec_token": "O F <space> C O U R S E <space> T H E R E <space> I S N ' T <space> A N Y <space> G U A R A N T E E <space> T H E <space> C O M P A N Y <space> W I L L <space> K E E P <space> I T S <space> T H O U G H <space> H A N D <eos>",
            "rec_tokenid": "34 25 18 22 34 40 37 38 24 18 39 27 24 37 24 18 28 38 33 5 39 18 20 33 44 18 26 40 20 37 20 33 39 24 24 18 39 27 24 18 22 34 32 35 20 33 44 18 42 28 31 31 18 30 24 24 35 18 28 39 38 18 39 27 34 40 26 27 18 27 20 33 23 51",
            "score": -70.69088745117188,
            "shape": [
                72,
                52
            ],
            "text": "OF COURSE THERE ISN'T ANY GUARANTEE THE COMPANY WILL KEEP ITS HOT HAND <NOISE>",
            "token": "O F <space> C O U R S E <space> T H E R E <space> I S N ' T <space> A N Y <space> G U A R A N T E E <space> T H E <space> C O M P A N Y <space> W I L L <space> K E E P <space> I T S <space> H O T <space> H A N D <space> <NOISE>",
            "tokenid": "34 25 18 22 34 40 37 38 24 18 39 27 24 37 24 18 28 38 33 5 39 18 20 33 44 18 26 40 20 37 20 33 39 24 24 18 39 27 24 18 22 34 32 35 20 33 44 18 42 28 31 31 18 30 24 24 35 18 28 39 38 18 27 34 39 18 27 20 33 23 18 17"
        }
    ],
```

**Comment by kan-bayashi** [comment]
@sw005320 Is it OK to merge?

**Pull request merged**